File: week7/hayeounglee/BOJ_폰 호석만.py
import sys
input = sys.stdin.readline

MAX = 2**63 - 1

A, B = input().split()

def char_to_val(c):
    if '0' <= c <= '9':
        return ord(c) - ord('0')
    return ord(c) - ord('a') + 10

def min_base(s):
    return max(char_to_val(c) for c in s) + 1

def convert(s, base):
    val = 0
    for c in s:
        v = char_to_val(c)
        if v >= base:
            return -1
        val = val * base + v
        if val > MAX:
            return -1
    return val

candidates = []

minA = min_base(A)
minB = min_base(B)

for baseA in range(minA, 37):
    valA = convert(A, baseA)
    if valA == -1:
        continue

    for baseB in range(minB, 37):
        if baseA == baseB:
            continue

        valB = convert(B, baseB)
        if valB == -1:
            continue

        if valA == valB:
            candidates.append((valA, baseA, baseB))

if len(candidates) == 0:
    print("Impossible")
elif len(candidates) > 1:
    print("Multiple")
else:
    x, a, b = candidates[0]
    print(x, a, b)

File: week7/yujin/BOJ_폰_호석만.js
const fs = require('fs');
const [XA, XB] = fs.readFileSync(0).toString().trim().split(' ');

function solution(XA, XB) {
  let answers = [];

  const getMinBase = str => {
    let maxVal = 0;
    for (let char of str) {
      const val = parseInt(char, 36);
      if (val > maxVal) maxVal = val;
    }
    return Math.max(2, maxVal + 1);
  };

  const minA = getMinBase(XA);
  const minB = getMinBase(XB);

  //모든 진법 조합 (2~36) 탐색
  for (let a = minA; a <= 36; a++) {
    for (let b = minB; b <= 36; b++) {
      if (a === b) continue;

      try {
        const valA = BigInt(parseInt(XA, a));
        const valB = BigInt(parseInt(XB, b));

        // 두 값이 같고 2^63 미만인지 확인
        if (valA === valB && valA < 2n ** 63n) {
          answers.push([valA.toString(), a, b]);
        }
      } catch (e) {
        // parseInt 과정에서 범위를 벗어나는 등 에러 발생 시 무시
        continue;
      }
    }
  }

  if (answers.length === 0) return 'Impossible';
  if (answers.length > 1) return 'Multiple';
  return answers[0].join(' ');
}

console.log(solution(XA, XB));
